### ISSUE TITLE
Question: Is there a reason why `Unbox a => Maybe (Unbox a)` doesn't exist?

### ISSUE THREAD

**Issue by LaurentRDC** [haskell/vector#455]
Hi,

I'm toying with the idea of implementing time-series analysis software on top of `vector`.

One key feature of other time-series analysis frameworks (e.g. python's `pandas`) is that there's a concept of missing values, or Not Available (NA).

I can already have:

```haskell
import Data.Vector
type VectorWithNA a = Vector (Maybe a) 
```

but I would like to take advantage of faster numerical operations of `Data.Vector.Unboxed`. However, I cannot do this:

```haskell
import Data.Vector.Unboxed (Unbox, Vector, fromList)

type UnboxedVectorWithNA a = Vector (Maybe a)

v :: UnboxedVectorWithNA Double
v = fromList [ Just 0] -- error: No instance for (Unbox (Maybe Double))
```

My question is: is this a fundamental limitation of the `Unbox` class?

**Comment by lehins** [comment]
Yes, there is a fundamental reason why. The only way you can unbox a `Maybe a` type is with two vectors, both of which have to be of the same length, matching the length of the original vector. Lets say you have this vector of `Maybe Int`

```haskell
v = [Just 1, Just 2, Nothing, Nothing, Just 5] :: Vector (Maybe Int)
```
In order to unbox it you have to use two vectors:
```haskell
v1 = [      1,       2,        _,         _,       5] :: Vector Int
v2 = [True, True, False, False, True] :: Vector Bool
```
This is necessary, because we do also have mutable vectors and there operations like unsafeThaw and unsafeFeeze that expect the memory layout to be the same between immutable and mutable vectors.

In other words if `v` is mutable then `v1` and `v2` are mutable and we should be able to modify those cells in place.

Unboxing is a technique for making things faster and reducing memory overhead. But if you compare this representation with a regular boxed vector, you don't get much improvement. So, there isn't really any gain in unboxing Maybe values.

Closing as duplicate of #119 

Feel free to reopen the ticket if you do come up with some strategy that is performant and safe :wink:
